### PR TITLE
Updating docs to clarify config precedence behavior

### LINKED ISSUE #2618: Mention different merge mechanisms for config fields
**Section/Content To Improve**
Section "Config order of precedence".

**Suggested Improvement**
The current implementation of `mergeConfig` use different merge mechanisms depends on different fields. Sometimes users felt confused, like #1000.

We should mention detailed fields in the document, or at least provide a link to the source code.

**Relevant File(s)**: README.md

### PR DESCRIPTION
**Pull request by leethomas**
Closed #2618. I made the following changes:
- enumerated the order of precedence description for a little more clarity
- added text describing the expected behavior (found this in the source code, let me know if I got something wrong)
- added an example illustrating the expected behavior

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -449,7 +449,7 @@ When using `catch`, or passing a [rejection callback](https://developer.mozilla.
 
 ## Config Defaults
 
-You can specify config defaults that will be applied to every request.
+You can specify config defaults that will be applied to every request, although there are some fields which will override your defaults if passed at the time of the request.  See [Config order of precedence](#config-order-of-precedence)
```
Comment (chinesedfan) [inline README.md]: See [Config order of precedence](#config-order-of-precedence) for more details about config merging strategies.

File: README.md [hunk 2]
```diff
@@ -490,6 +497,22 @@ instance.get('/longRequest', {
...
+// Override the params default for the library
+// Now all requests using this instance will pass the param defined below
+instance.defaults.params = { q: 'some-value' }
+
+// The `params` object will be entirely overridden, and the
+// previously defined key `q` will no longer exist
+instance.get('/testRequest', {
+  params: { a: 'a-new-value' }
+})
+```
```
Comment (chinesedfan) [inline README.md]: Thanks for your sweet examples. But I am afraid that `params` will be changed with #2190.

File: README.md [hunk 3]
```diff
@@ -473,7 +473,14 @@ instance.defaults.headers.common['Authorization'] = AUTH_TOKEN;
 
 ### Config order of precedence
 
-Config will be merged with an order of precedence. The order is library defaults found in [lib/defaults.js](https://github.com/axios/axios/blob/master/lib/defaults.js#L28), then `defaults` property of the instance, and finally `config` argument for the request. The latter will take precedence over the former. Here's an example.
+The config will be merged with the following order of precedence:
+1. Library defaults found in [lib/defaults.js](https://github.com/axios/axios/blob/master/lib/defaults.js#L28)
+2. The `defaults` property of the instance
+3. The `config` argument for the request.
```
Comment (chinesedfan) [inline README.md]: The old paragraph is fine. Unless there is something wrong, I will leave it as it is.

File: README.md [hunk 4]
```diff
@@ -473,7 +473,14 @@ instance.defaults.headers.common['Authorization'] = AUTH_TOKEN;
...
+
+Passing the `config` argument will completely **override** what is specified in the `defaults` property for some keys. These are currently `url`, `method`, `params`, and `data`.
```
Comment (chinesedfan) [inline README.md]: That's what I want to emphasize for users. We have different merging strategies for different fields, i.e.

- always from request config, i.e. url/method
- deep merge, i.e. headers
- merge by replacement, i.e. other fields

And I also want to mention something that will affect the final config. Including but no limiting for,

- relationships between config.header.common/[method] and config[header]
- the role of default `config.transformRequest`
- interceptors

**Comment by jasonsaayman** [comment]
Hi @leethomas,

Axios docs have now been moved to a new repo [https://github.com/axios/axios-docs](https://github.com/axios/axios-docs). We have decided on a new repo with a website that is built to handle this.

Thanks

**Pull request closed without merging**